Tools: Insert a CSV into the docs

codeallthethingz opened the pull request:
This PR adds the ability to inline a CSV into a markdown document through the github_readme_sync tool.

Using the syntax

```
!table[../../data.csv]
```

The sync tool will check if `data.csv` exists and if it does, it will convert the CSV into an HTML table.

The CSV header row can contain optional formatting parameters using a bar separated syntax (in any order)
- `Correct (%)|hover Number of correct...`, the part after the pipe is extracted, escaped and put into a title attribute on that header row in the table.  See the image below.
- `Correct (%)|align {right,left}|hover ...` the alignment is changed to right or left for the data cells. If no alignment is specified, there is no alignment added to the HTML and the alignment will follow any CSS rules that exist on the page.

![image](https://github.com/user-attachments/assets/9ae4e1e2-a5f2-4934-a9bd-3916a4cfc75f)

This can be previewed here on the benchmarks page https://thousandbrainsproject.readme.io/v0.0-table-test/docs/benchmark-experiments#results

> [!WARNING]
> the auto generated PR docs preview system will not be correct as it uses code from the main branch.

codeallthethingz commented on tools/github_readme_sync/hierarchy.py:
```diff
@@ -170,9 +180,7 @@ def check_links(path):
-        # Remove any #hash fragments from the path
-        path = match.split("#")[0]
-        path_to_check = os.path.join(current_dir, path)
+        path_to_check = os.path.join(current_dir, match.split("#")[0])
```
It's a bug that needed to be fixed.  The path being overridden was causing the wrong filename to be printed out in the error string below.

tristanls replied:
I trust that you had reasons for updating this, as you shared. I noted that this is a separate pull request that can be merged after 30 seconds of review rather than being attached to this pull request.

codeallthethingz commented on tools/github_readme_sync/hierarchy.py:
```diff
@@ -255,7 +263,7 @@ def read_file_content(file_path):
 
 
-def extract_links(content):
+def extract_external_links(content):
```
While making the changes I was confused by this method name and spent time editing it until I realized it was only for external links.  I updated to prevent future Will from doing the same thing.

codeallthethingz commented on tools/github_readme_sync/readme.py:
```diff
@@ -168,10 +174,78 @@ def create_category_if_not_exists(self, slug: str, title: str) -> Tuple[str, boo
+                    for col_idx in range(len(headers)):
+                        is_numeric_col = True
+                        for row in rows:
+                            if not row[col_idx].replace(".", "").isdigit():
```
I found pint - and applied it to this problem. https://pint.readthedocs.io/en/stable/

tristanls replied:
NOICE!

**chore**: The `pint` dependency needs to be added to `pyproject.toml`.

**issue**: I think using `ureg.parse_expression(s)` will claim that some strings that don't meet the definition expected here are numeric. The absence of a throw in my paste below means `is_numeric()` will return `True` in all cases below.
```
>>> from pint import UnitRegistry
>>> ureg = UnitRegistry()
>>> ureg.parse_expression("-1")
-1
>>> ureg.parse_expression("3/4")
0.75
>>> ureg.parse_expression("10e4")
100000.0
>>> ureg.parse_expression("12.83")
12.83
>>> ureg.parse_expression("1 000")
0
>>> ureg.parse_expression("1 000 000")
0
>>> ureg.parse_expression("10s")
<Quantity(10, 'second')>
>>> ureg.parse_expression("1,000")
1000
>>> ureg.parse_expression("10 meters")
<Quantity(10, 'meter')>
```
**question**: Why does `10s` or `1,000` return `False` in your code paste?

codeallthethingz replied:
added some more unit tests and fixed this.

codeallthethingz commented on tools/github_readme_sync/readme.py:
```diff
@@ -168,10 +174,78 @@ def create_category_if_not_exists(self, slug: str, title: str) -> Tuple[str, boo
+                        title_attr = ""
+                        title_parts = header.split("|")
+                        if len(title_parts) > 1:
+                            title = html.escape(title_parts[1], quote=True)
```
I did add `nh3.clean`.  I still had to escape the title before it goes into nh3 as it just removes the quoted section rather than escaping the quotes.

tristanls commented on benchmarks/results/ycb_10objs.csv:
```diff
@@ -1,13 +1,13 @@
+base_10simobj_surf_agent,94.29,10.71,81,13.75,10,50
+randrot_noise_10simobj_dist_agent,84.00,40.00,237,34.95,18,147
+randrot_noise_10simobj_surf_agent,93.00,33.00,176,29.79,24,205
+randomrot_rawnoise_10distinctobj_surf_agent,68.00,78.00,16,102.01,17,15
```
**issue**: The conversion from radians to degrees looks incorrect.

```
1.78 * 180 / π = 320.4 / π = 101.98648753 ~> 101.99
```
> [!NOTE]
> I only spot-checked this one; likely, all others are incorrect if they used the same formula.

tristanls commented on tools/github_readme_sync/readme.py:
```diff
@@ -168,10 +174,78 @@ def create_category_if_not_exists(self, slug: str, title: str) -> Tuple[str, boo
+                        table += "</tr>\n"
+
+                    table += "</tbody>\n</table></div>"
+                    return table
```
Thanks!

tristanls commented on benchmarks/results/ycb_10objs.csv:
```diff
@@ -1,13 +1,13 @@
+Experiment,Correct (%),Used MLH (%),Num Match Steps,Rotation Error (degrees),Run Time (mins),Episode Run Time (s)
+base_config_10distinctobj_dist_agent,100.00,0.00,37,16.04,5,19
+base_config_10distinctobj_surf_agent,100.00,0.00,28,13.18,4,18
+randrot_noise_10distinctobj_dist_agent,100.00,3.00,55,22.34,5,33
```
22.34535
```suggestion
randrot_noise_10distinctobj_dist_agent,100.00,3.00,55,22.35,5,33
```

tristanls commented on benchmarks/results/ycb_10objs.csv:
```diff
@@ -1,13 +1,13 @@
+randrot_noise_10distinctobj_dist_on_distm,99.00,1.00,38,12.03,4,28
+randrot_noise_10distinctobj_surf_agent,100.00,0.00,29,22.34,5,33
+randrot_10distinctobj_surf_agent,100.00,1.00,28,18.33,4,21
+randrot_noise_10distinctobj_5lms_dist_agent,100.00,0.00,57,45.27,10,84
```
45.26367

```suggestion
randrot_noise_10distinctobj_5lms_dist_agent,100.00,0.00,57,45.26,10,84
```

tristanls commented on benchmarks/results/ycb_77objs.csv:
```diff
@@ -1,6 +1,6 @@
-randrot_noise_77obj_5lms_dist_agent,88.31%,0.00%,70,1.02,32m,862s
\ No newline at end of file
+Experiment,Correct (%),Used MLH (%),Num Match Steps,Rotation Error (degrees),Run Time (mins),Episode Run Time (s)
+base_77obj_dist_agent,93.51,12.99,108,0.28,62,212
```
```suggestion
base_77obj_dist_agent,93.51,12.99,108,16.04,62,212
```

tristanls commented on benchmarks/results/ycb_77objs.csv:
```diff
@@ -1,6 +1,6 @@
\ No newline at end of file
+Experiment,Correct (%),Used MLH (%),Num Match Steps,Rotation Error (degrees),Run Time (mins),Episode Run Time (s)
+base_77obj_dist_agent,93.51,12.99,108,0.28,62,212
+base_77obj_surf_agent,99.13,6.06,54,0.19,28,82
```
```suggestion
base_77obj_surf_agent,99.13,6.06,54,10.89,28,82
```

tristanls commented on benchmarks/results/ycb_77objs.csv:
```diff
@@ -1,6 +1,6 @@
+Experiment,Correct (%),Used MLH (%),Num Match Steps,Rotation Error (degrees),Run Time (mins),Episode Run Time (s)
+base_77obj_dist_agent,93.51,12.99,108,0.28,62,212
+base_77obj_surf_agent,99.13,6.06,54,0.19,28,82
+randrot_noise_77obj_dist_agent,89.61,22.51,152,0.65,85,308
```
```suggestion
randrot_noise_77obj_dist_agent,89.61,22.51,152,37.24,85,308
```

tristanls commented on benchmarks/results/ycb_77objs.csv:
```diff
@@ -1,6 +1,6 @@
+base_77obj_dist_agent,93.51,12.99,108,0.28,62,212
+base_77obj_surf_agent,99.13,6.06,54,0.19,28,82
+randrot_noise_77obj_dist_agent,89.61,22.51,152,0.65,85,308
+randrot_noise_77obj_surf_agent,92.64,25.11,120,0.69,65,227
```
```suggestion
randrot_noise_77obj_surf_agent,92.64,25.11,120,39.53,65,227
```

tristanls commented on benchmarks/results/ycb_77objs.csv:
```diff
@@ -1,6 +1,6 @@
+base_77obj_surf_agent,99.13,6.06,54,0.19,28,82
+randrot_noise_77obj_dist_agent,89.61,22.51,152,0.65,85,308
+randrot_noise_77obj_surf_agent,92.64,25.11,120,0.69,65,227
+randrot_noise_77obj_5lms_dist_agent,88.31,0.00,70,1.02,32,862
```
```suggestion
randrot_noise_77obj_5lms_dist_agent,88.31,0.00,70,58.44,32,862
```

tristanls reviewed (commented):
I checked all radian to degree conversions. Suggested updates.

codeallthethingz commented:
@tristanls quick update as there were structural changes.  

- I removed the intelligent alignment of columns based on the data as this was mixing display and data logic, and moved that logic into optional syntax in the CSV header row.  See the PR description for details.

tristanls commented:
FYI, I missed this being marked ready for review. I don't think GitHub sends a notification for that state change. Taking a look now.

tristanls reviewed (approved):
It looks good. Thanks for hanging in there. I think this is a good approach. I believe one function needs to be moved, but things look good otherwise.
